File: frontend/lib/services/api_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import '../config/app_config.dart';
import 'storage_service.dart';

// Custom exceptions
class ApiException implements Exception {
  final String message;
  final int? statusCode;
  ApiException(this.message, {this.statusCode});
  @override
  String toString() => message;
}

class ValidationException implements Exception {
  final String message;
  final Map<String, List<String>> errors;
  ValidationException(this.message, this.errors);

  String get firstError {
    if (errors.isEmpty) return message;
    return errors.values.first.first;
  }

  @override
  String toString() => firstError;
}

class NetworkException implements Exception {
  final String message;
  NetworkException(
      [this.message = 'Network error. Please check your connection.']);
  @override
  String toString() => message;
}

class UnauthorisedException implements Exception {
  final String message;
  UnauthorisedException(
      [this.message = 'Session expired. Please login again.']);
  @override
  String toString() => message;
}

// Navigation callback for 401 redirect
typedef OnUnauthorised = void Function();

class ApiService {
  late final Dio _dio;
  OnUnauthorised? onUnauthorised;

  ApiService() {
    _dio = Dio(
      BaseOptions(
        baseUrl: AppConfig.baseUrl,
        connectTimeout: Duration(seconds: AppConfig.connectTimeoutSeconds),
        receiveTimeout: Duration(seconds: AppConfig.receiveTimeoutSeconds),
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
      ),
    );

    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) async {
          final token = await StorageService.getToken();
          if (token != null) {
            options.headers['Authorization'] = 'Bearer $token';
          }
          if (kDebugMode) {
            debugPrint('→ ${options.method} ${options.path}');
            if (options.data != null) debugPrint('   Body: ${options.data}');
          }
          handler.next(options);
        },
        onResponse: (response, handler) {
          if (kDebugMode) {
            debugPrint(
                '← ${response.statusCode} ${response.requestOptions.path}');
          }
          handler.next(response);
        },
        onError: (error, handler) async {
          if (kDebugMode) {
            debugPrint(
                '✗ ${error.response?.statusCode} ${error.requestOptions.path}: ${error.message}');
          }

          final statusCode = error.response?.statusCode;

          if (statusCode == 401) {
            await StorageService.clearAll();
            onUnauthorised?.call();
            handler.reject(DioException(
              requestOptions: error.requestOptions,
              error: UnauthorisedException(),
              type: DioExceptionType.badResponse,
            ));
            return;
          }

          if (statusCode == 422) {
            final data = error.response?.data;
            String message = 'Validation failed';
            Map<String, List<String>> errors = {};
            if (data is Map) {
              message = data['message'] as String? ?? message;
              final rawErrors = data['errors'];
              if (rawErrors is Map) {
                errors = rawErrors.map((k, v) {
                  final list = v is List
                      ? v.map((e) => e.toString()).toList()
                      : [v.toString()];
                  return MapEntry(k.toString(), list);
                });
              }
            }
            handler.reject(DioException(
              requestOptions: error.requestOptions,
              error: ValidationException(message, errors),
              type: DioExceptionType.badResponse,
            ));
            return;
          }

          if (error.type == DioExceptionType.connectionTimeout ||
              error.type == DioExceptionType.receiveTimeout ||
              error.type == DioExceptionType.connectionError) {
            handler.reject(DioException(
              requestOptions: error.requestOptions,
              error: NetworkException(),
              type: error.type,
            ));
            return;
          }

          handler.next(error);
        },
      ),
    );
  }

  /// Unwrap the exception from DioException
  static Exception _unwrap(dynamic e) {
    if (e is DioException && e.error != null) {
      final inner = e.error;
      if (inner is ApiException) return inner;
      if (inner is ValidationException) return inner;
      if (inner is NetworkException) return inner;
      if (inner is UnauthorisedException) return inner;
    }
    if (e is ApiException) return e;
    if (e is ValidationException) return e;
    if (e is NetworkException) return e;
    if (e is UnauthorisedException) return e;
    final statusCode = (e is DioException) ? e.response?.statusCode : null;
    final message = (e is DioException)
        ? (e.response?.data?['message'] as String? ??
            e.message ??
            'Unknown error')
        : e.toString();
    return ApiException(message, statusCode: statusCode);
  }

  Future<dynamic> get(String path,
      {Map<String, dynamic>? queryParameters}) async {
    try {
      final response = await _dio.get(path, queryParameters: queryParameters);
      return response.data;
    } catch (e) {
      throw _unwrap(e);
    }
  }

  Future<dynamic> post(String path, {dynamic data}) async {
    try {
      final response = await _dio.post(path, data: data);
      return response.data;
    } catch (e) {
      throw _unwrap(e);
    }
  }

  Future<dynamic> put(String path, {dynamic data}) async {
    try {
      final response = await _dio.put(path, data: data);
      return response.data;
    } catch (e) {
      throw _unwrap(e);
    }
  }

  Future<dynamic> delete(String path) async {
    try {
      final response = await _dio.delete(path);
      return response.data;
    } catch (e) {
      throw _unwrap(e);
    }
  }

  /// For PDF downloads — returns raw bytes
  Future<List<int>> getBytes(String path,
      {Map<String, dynamic>? queryParameters}) async {
    try {
      final response = await _dio.get<List<int>>(
        path,
        queryParameters: queryParameters,
        options: Options(responseType: ResponseType.bytes),
      );
      return response.data ?? [];
    } catch (e) {
      throw _unwrap(e);
    }
  }
}

File: ios/Runner/Services/APIService.swift
import Foundation

// MARK: - Error types

enum APIError: LocalizedError {
    case network(String)
    case unauthorised
    case validation(String, [String: [String]])
    case server(Int, String)
    case decoding(String)
    case unknown(String)

    var errorDescription: String? {
        switch self {
        case .network(let m):         return m
        case .unauthorised:           return "Session expired. Please login again."
        case .validation(_, let e):
            return e.values.first?.first ?? "Validation failed."
        case .server(_, let m):       return m
        case .decoding(let m):        return "Decoding error: \(m)"
        case .unknown(let m):         return m
        }
    }
}

// MARK: - Pagination wrapper

struct PaginatedResponse<T: Decodable>: Decodable {
    let data: [T]
    let meta: Meta?

    struct Meta: Decodable {
        let currentPage: Int?
        let lastPage: Int?
        let total: Int?
        let totalAmount: Double?

        enum CodingKeys: String, CodingKey {
            case currentPage  = "current_page"
            case lastPage     = "last_page"
            case total
            case totalAmount  = "total_amount"
        }
    }
}

// MARK: - Callback for 401

typealias UnauthorisedHandler = @MainActor () -> Void

// MARK: - API Service

@MainActor
final class APIService: ObservableObject {
    static let shared = APIService()
    var onUnauthorised: UnauthorisedHandler?

    private let session: URLSession

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest  = AppConfig.connectTimeoutSeconds
        config.timeoutIntervalForResource = AppConfig.receiveTimeoutSeconds
        config.httpAdditionalHeaders = [
            "Accept":       "application/json",
            "Content-Type": "application/json"
        ]
        session = URLSession(configuration: config)
    }

    // MARK: HTTP verbs

    func get<T: Decodable>(_ path: String, query: [String: Any] = [:]) async throws -> T {
        try await request(method: "GET", path: path, query: query, body: nil)
    }

    func post<T: Decodable>(_ path: String, body: [String: Any]? = nil) async throws -> T {
        try await request(method: "POST", path: path, query: [:], body: body)
    }

    func put<T: Decodable>(_ path: String, body: [String: Any]? = nil) async throws -> T {
        try await request(method: "PUT", path: path, query: [:], body: body)
    }

    func patch<T: Decodable>(_ path: String, body: [String: Any]? = nil) async throws -> T {
        try await request(method: "PATCH", path: path, query: [:], body: body)
    }

    func delete<T: Decodable>(_ path: String) async throws -> T {
        try await request(method: "DELETE", path: path, query: [:], body: nil)
    }

    /// Fire-and-forget version (ignores response body)
    func postVoid(_ path: String, body: [String: Any]? = nil) async throws {
        let _: EmptyResponse = try await request(method: "POST", path: path, query: [:], body: body)
    }

    // MARK: - Private

    private func request<T: Decodable>(
        method: String,
        path: String,
        query: [String: Any],
        body: [String: Any]?
    ) async throws -> T {
        var components = URLComponents(string: AppConfig.baseURL + path)!
        if !query.isEmpty {
            components.queryItems = query.map {
                URLQueryItem(name: $0.key, value: "\($0.value)")
            }
        }
        var req = URLRequest(url: components.url!)
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token = try? KeychainService.read(for: AppConfig.tokenKey) {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let body {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        #if DEBUG
        print("→ \(method) \(path)")
        #endif

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: req)
        } catch {
            throw APIError.network("Network error. Please check your connection.")
        }

        guard let http = response as? HTTPURLResponse else {
            throw APIError.unknown("Invalid response")
        }

        #if DEBUG
        print("← \(http.statusCode) \(path)")
        #endif

        switch http.statusCode {
        case 200...299:
            if T.self == EmptyResponse.self {
                return EmptyResponse() as! T
            }
            do {
                return try JSONDecoder().decode(T.self, from: data)
            } catch {
                throw APIError.decoding(error.localizedDescription)
            }
        case 401:
            KeychainService.delete(for: AppConfig.tokenKey)
            KeychainService.delete(for: AppConfig.userKey)
            onUnauthorised?()
            throw APIError.unauthorised
        case 422:
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
                let msg = json["message"] as? String ?? "Validation failed"
                var errors: [String: [String]] = [:]
                if let rawErrors = json["errors"] as? [String: Any] {
                    for (k, v) in rawErrors {
                        if let arr = v as? [String] {
                            errors[k] = arr
                        } else {
                            errors[k] = ["\(v)"]
                        }
                    }
                }
                throw APIError.validation(msg, errors)
            }
            throw APIError.server(422, "Validation failed")
        default:
            let msg = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["message"] as? String
                ?? HTTPURLResponse.localizedString(forStatusCode: http.statusCode)
            throw APIError.server(http.statusCode, msg)
        }
    }
}

private struct EmptyResponse: Decodable {}
